### PR TITLE
Adding Functionality to Find or Create Users and Add Them to Cohorts, Changed Sample Data File

### PR DESCRIPTION
**Pull request by brvarner**
This is a big update, it:
- Adds cohort member creation form
- Add cohort member show page
- Adds a before_action to the cohort_members controller that finds or creates the user that the admin is attempting to add to a cohort
  - Gives the user a randomly generated password as well
- Changes CohortMember and User model associations to use email as the primary key
- Adds a migration to change the User ID field to an email field in the Cohort_Members Table

### CODE REVIEW

File: app/models/cohort_member.rb [hunk 1]
```diff
@@ -3,15 +3,14 @@
...
 #  role       :string
 #  created_at :datetime         not null
 #  updated_at :datetime         not null
 #  capacity   :integer
 #
 class CohortMember < ApplicationRecord
-  belongs_to :user, required: true, class_name: "User", foreign_key: "user_id"
-  belongs_to :cohort, required: true, class_name: "Cohort", foreign_key: "cohort_id"
-
+  belongs_to :user, required: true, class_name: 'User', primary_key: 'email', foreign_key: 'email'
```
Comment (MayaS1111) [inline app/models/cohort_member.rb:14]: good

File: app/models/user.rb [hunk 2]
```diff
@@ -28,52 +28,64 @@ class User < ApplicationRecord
...
-  has_many :cohort_members, class_name: "CohortMember", foreign_key: "user_id", dependent: :destroy
-  has_one :mentor, class_name: "Match", foreign_key: "mentor_id", dependent: :destroy
-  has_many :mentees, class_name: "Match", foreign_key: "mentee_id", dependent: :destroy
-  has_many :mentor_submissions, class_name: "MatchSubmission", foreign_key: "mentor_id", dependent: :destroy
-  has_many :mentee_submissions, class_name: "MatchSubmission", foreign_key: "mentee_id", dependent: :destroy
-  has_many :owned_cohorts, class_name: "Cohort", foreign_key: "creator_id", dependent: :destroy
-  has_many :owned_programs, class_name: "Program", foreign_key: "creator_id", dependent: :destroy
-  
-  
+  has_many :cohort_members, class_name: 'CohortMember', primary_key: 'email', foreign_key: 'email', dependent: :destroy
```
Comment (MayaS1111) [inline app/models/user.rb:31]: good

File: app/controllers/cohort_members_controller.rb [hunk 3]
```diff
@@ -52,19 +53,29 @@ def destroy
...
+  def set_cohort_member
+    @cohort_member = CohortMember.find(params[:id])
+  end
+
+  # Only allow a list of trusted parameters through.
+  def cohort_member_params
+    params.require(:cohort_member).permit(:email, :cohort_id, :role)
+  end
+
+  def find_or_create_user(cohort_member_params)
```
Comment (MayaS1111) [inline app/controllers/cohort_members_controller.rb:73]: I like how you did this

File: app/models/user.rb [hunk 4]
```diff
@@ -28,52 +28,64 @@ class User < ApplicationRecord
...
   # Returns list of mentees that are in the same cohort as the provided mentor
-  scope :mentors_in_cohort, ->(cohort) { joins(:cohort_members)
-                                  .where('cohort_members.cohort_id = ? AND cohort_members.role = ?', cohort, 'mentor')}
-  scope :mentees_in_cohort, ->(cohort) { joins(:cohort_members)
-                                  .where('cohort_members.cohort_id = ? AND cohort_members.role = ?', cohort, 'mentee')}
-  scope :unpaired_mentees_in_cohort, ->(cohort) { joins(:cohort_members)
-                                  .left_joins('LEFT JOIN matches ON matches.mentee_id = users.id AND matches.active = true')
-                                  .where('cohort_members.cohort_id = ? AND cohort_members.role = ?', cohort, 'mentee')
-                                  .where('matches.id IS NULL')}
+  scope :mentors_in_cohort, lambda { |cohort|
```
Comment (MayaS1111) [inline app/models/user.rb:46]: Can you change this back:
I like the other way best " -> (cohort) {....."

File: app/models/user.rb [hunk 5]
```diff
@@ -28,52 +28,64 @@ class User < ApplicationRecord
...
+                                         .where('cohort_members.cohort_id = ? AND cohort_members.role = ?', cohort, 'mentee')
+                                         .where('matches.id IS NULL')
+                                     }
   def name
     "#{first_name.capitalize} #{last_name.capitalize}"
   end
+
   def matched?
-    Match.where('mentee_id = :id OR mentor_id = :id', id: self.id).exists?
+    Match.where('mentee_id = :id OR mentor_id = :id', id:).exists?
```
Comment (MayaS1111) [inline app/models/user.rb:65]: i like!

File: app/models/user.rb [hunk 6]
```diff
@@ -28,52 +28,64 @@ class User < ApplicationRecord
...
   end
+
   def cohort
-    cohort_id = CohortMember.where(user_id: self.id).pluck(:cohort_id).first
+    cohort_id = CohortMember.where(email:).pluck(:cohort_id).first
```
Comment (MayaS1111) [inline app/models/user.rb:69]: YES, I forgot these would need to be updated too!

File: lib/tasks/dev.rake [hunk 7]
```diff
@@ -28,32 +28,45 @@ task({ sample_data: :environment }) do
...
-    mentor_title = ['Software Engineer', 'Consultant', 'Technical Support', 'IT Technician', 'Project Manager', 'Product Manager', 'UI/UX Designer', 'Sales Coordinator'].sample
-    inactive_reason = ['Did not like the platform.', 'Did not have a good experience.', 'I will be back!', 'Other'].sample
-    role = { 'admin' => 5, 'observer' => 7, 'mentor' => 25, 'mentee' => 100 }.find { |_key, value| rand * 100 <= value }.first
+    timezone = ['Eastern Standard Time (EST) - UTC-5', 'Central Standard Time (CST) - UTC-6',
+                'Pacific Standard Time (PST) - UTC-8', 'Mountain Standard Time (MST) - UTC-7'].sample
+    mentor_title = ['Software Engineer', 'Consultant', 'Technical Support', 'IT Technician', 'Project Manager',
+                    'Product Manager', 'UI/UX Designer', 'Sales Coordinator'].sample
+    inactive_reason = ['Did not like the platform.', 'Did not have a good experience.', 'I will be back!',
+                       'Other'].sample
+    role = { 'admin' => 5, 'observer' => 7, 'mentor' => 25, 'mentee' => 100 }.find do |_key, value|
```
Comment (MayaS1111) [inline lib/tasks/dev.rake:40]: Can you change this role line back?

File: lib/tasks/dev.rake [hunk 8]
```diff
@@ -138,22 +157,27 @@ task({ sample_data: :environment }) do
...
       cohort_id: cohort.id,
-      role: role,
-      capacity: capacity,
+      role:,
+      capacity:
     )
   end
 
-# Creating Matches
+  # Creating Matches
```
Comment (MayaS1111) [inline lib/tasks/dev.rake:167]: Thanks for updating this section!

**Review by MayaS1111 — CHANGES_REQUESTED** [review]
Looks great!

Are you able to change 2 parts back:
- The scopes in the user model (" -> (cohort)")
- The role in that sample data (line 40)

File: app/controllers/cohort_members_controller.rb
Comment (APatel-AI) [inline app/controllers/cohort_members_controller.rb:1]: Controller implementation looks solid. The permissions and logic are well-structured and align with its intended purpose.

File: app/models/cohort_member.rb
Comment (APatel-AI) [inline app/models/cohort_member.rb:1]: Looks good!

**Review by APatel-AI — APPROVED** [review]
Looks great!

File: app/models/user.rb [hunk 9]
```diff
@@ -28,52 +28,59 @@ class User < ApplicationRecord
...
   after_create :create_first_cohort
   attr_accessor :cohort_members_attributes
+
   before_create :set_default_active_status
 
   # Returns list of mentees that are in the same cohort as the provided mentor
-  scope :mentors_in_cohort, ->(cohort) { joins(:cohort_members)
+ scope :mentors_in_cohort, ->(cohort) { joins(:cohort_members)
                                   .where('cohort_members.cohort_id = ? AND cohort_members.role = ?', cohort, 'mentor')}
   scope :mentees_in_cohort, ->(cohort) { joins(:cohort_members)
```
Comment (MayaS1111) [inline app/models/user.rb:48]: Great thanks!

File: lib/tasks/dev.rake [hunk 10]
```diff
@@ -28,32 +28,44 @@ task({ sample_data: :environment }) do
...
-    timezone = ['Eastern Standard Time (EST) - UTC-5', 'Central Standard Time (CST) - UTC-6', 'Pacific Standard Time (PST) - UTC-8', 'Mountain Standard Time (MST) - UTC-7'].sample
-    mentor_title = ['Software Engineer', 'Consultant', 'Technical Support', 'IT Technician', 'Project Manager', 'Product Manager', 'UI/UX Designer', 'Sales Coordinator'].sample
-    inactive_reason = ['Did not like the platform.', 'Did not have a good experience.', 'I will be back!', 'Other'].sample
+    timezone = ['Eastern Standard Time (EST) - UTC-5', 'Central Standard Time (CST) - UTC-6',
+                'Pacific Standard Time (PST) - UTC-8', 'Mountain Standard Time (MST) - UTC-7'].sample
+    mentor_title = ['Software Engineer', 'Consultant', 'Technical Support', 'IT Technician', 'Project Manager',
+                    'Product Manager', 'UI/UX Designer', 'Sales Coordinator'].sample
+    inactive_reason = ['Did not like the platform.', 'Did not have a good experience.', 'I will be back!',
+                       'Other'].sample
     role = { 'admin' => 5, 'observer' => 7, 'mentor' => 25, 'mentee' => 100 }.find { |_key, value| rand * 100 <= value }.first
```
Comment (MayaS1111) [inline lib/tasks/dev.rake:40]: Thanks

**Review by MayaS1111 — APPROVED** [review]
Thanks for the changes! Amazing work Brandon

**Pull request merged**